### PR TITLE
Added string_stream

### PR DESCRIPTION
**Pull request by juliusnehring**

### CODE REVIEW

File: src/clean-core/string_stream.hh [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+                m_data = new char[rhs.size()];
+                m_capacity = rhs.size();
+            }
+            std::memcpy(m_data, rhs.m_data, rhs.size());
+            m_curr = m_data + rhs.size();
+        }
+        return *this;
+    };
+
+    string_stream& operator=(string_stream&& rhs) noexcept
```
Comment (jkunstwald) [inline src/clean-core/string_stream.hh:96]: possibly have to add a check to prevent self-move

File: src/clean-core/string_stream.hh [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    [[nodiscard]] size_t size() const noexcept { return m_curr - m_data; }
+    [[nodiscard]] bool empty() const { return m_curr == m_data; }
+
+public: // ctor
+    string_stream() = default;
+
+    string_stream(string_stream const& rhs)
+    {
+        m_data = new char[rhs.size()];
+        std::memcpy(m_data, rhs.m_data, rhs.size());
```
Comment (jkunstwald) [inline src/clean-core/string_stream.hh]: this is UB if `rhs` is empty

File: src/clean-core/string_stream.hh [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+public: // assignment
+    string_stream& operator=(string_stream const& rhs)
+    {
+        if (this != &rhs)
+        {
+            if (m_capacity < rhs.size())
+            {
+                delete[] m_data;
+                m_data = new char[rhs.size()];
```
Comment (jkunstwald) [inline src/clean-core/string_stream.hh:80]: this behaves unintuitive if `rhs` is empty (size will be 0 but m_data won't be `nullptr`)

File: src/clean-core/string_stream.hh [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+public: // properties
+    [[nodiscard]] size_t size() const noexcept { return m_curr - m_data; }
+    [[nodiscard]] bool empty() const { return m_curr == m_data; }
+
+public: // ctor
+    string_stream() = default;
+
+    string_stream(string_stream const& rhs)
+    {
+        m_data = new char[rhs.size()];
```
Comment (jkunstwald) [inline src/clean-core/string_stream.hh]: this behaves unintuitively if `rhs` is empty (size 0, but data not `nullptr`)

File: src/clean-core/string_stream.hh [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    /// reseve space for at least size more elements
+    void reserve(size_t size)
+    {
+        size_t req_size = this->size() + size;
+        if (req_size <= m_capacity)
+            return;
+
+        size_t new_cap = (m_capacity << 1) < req_size ? req_size : (m_capacity << 1);
+        char* new_data = new char[new_cap];
+        std::memcpy(new_data, m_data, this->size());
```
Comment (jkunstwald) [inline src/clean-core/string_stream.hh]: this is UB if the stringstream is default-initialized

Comment (Philip-Trettner) [reply src/clean-core/string_stream.hh]: good point.

> If either dest or src is an invalid or null pointer, the behavior is undefined, even if count is zero.

File: src/clean-core/string_stream.hh (hunk 1 above)
Comment (Philip-Trettner) [reply src/clean-core/string_stream.hh:96]: that is actually ok here. self-move will leave the string_stream in a proper empty state.

File: src/clean-core/string_stream.hh [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        return *this;
+    };
+
+private: // member
+    char* m_data = nullptr;
+    char* m_curr = nullptr;
+    size_t m_capacity = 0;
+};
+
+string to_string(string_stream const& ss) { return ss.to_string(); }
```
Comment (Philip-Trettner) [inline src/clean-core/string_stream.hh]: missing `inline`

File: src/clean-core/string_stream.hh [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        reserve(sv.size());
+        std::memcpy(m_curr, sv.data(), sv.size());
+        m_curr += sv.size();
+        return *this;
+    }
+
+    [[nodiscard]] string to_string() const
+    {
+        string s = string::uninitialized(size());
+        memcpy(s.data(), m_data, size());
```
Comment (Philip-Trettner) [inline src/clean-core/string_stream.hh]: `std::` for consistency

File: src/clean-core/string_stream.hh (hunk 1 above)
Comment (jkunstwald) [reply src/clean-core/string_stream.hh:96]: right!

File: src/clean-core/string_stream.hh [hunk 8]
```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+#include <clean-core/string.hh>
+#include <clean-core/string_view.hh>
+#include <cstring> // std::memcpy
```
Comment (jkunstwald) [inline src/clean-core/string_stream.hh]: please include STL headers in a block before the cc ones

File: src/clean-core/string_stream.hh [hunk 9]
```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+#include <clean-core/string.hh>
+#include <clean-core/string_view.hh>
```
Comment (Philip-Trettner) [inline src/clean-core/string_stream.hh:4]: I wonder if we should at least only use a forward decl for string

**Review by Philip-Trettner — APPROVED** [review]

File: src/clean-core/string_stream.hh [hunk 10]
```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include <cstring> // std::memcpy
+
+#include <clean-core/string.hh>
+#include <clean-core/string_view.hh>
+
+namespace cc
+{
+class string_stream
```
Comment (Philip-Trettner) [inline src/clean-core/string_stream.hh]: ah just saw: in cc we use `struct` for value types like these. Also please add string_stream to `fwd.hh`

**Review by Philip-Trettner — APPROVED** [review]

**Pull request merged**